examples/so_vits_svc.py

```diff
@@ -0,0 +1,834 @@
+import sys, os, logging, time, io, math, argparse, numpy as np
+from pathlib import Path
+from typing import Tuple, Optional
+from tinygrad import nn
+from tinygrad.tensor import Tensor
+from tinygrad.helpers import dtypes, getenv
+from tinygrad.state import torch_load
+from examples.vits import ResidualCouplingBlock, PosteriorEncoder, Encoder, ResBlock1, ResBlock2, LRELU_SLOPE, sequence_mask, split, download_if_not_present, get_hparams_from_file, load_checkpoint, weight_norm, HParams
+from examples.sovits_helpers import preprocess
+import soundfile
+
+# original code: https://github.com/svc-develop-team/so-vits-svc
+
+DEBUG = getenv("DEBUG")
+
+F0_BIN = 256
+F0_MAX = 1100.0
+F0_MIN = 50.0
+F0_MEL_MIN = 1127 * np.log(1 + F0_MIN / 700)
+F0_MEL_MAX = 1127 * np.log(1 + F0_MAX / 700)
+
+# original code for contentvec: https://github.com/auspicious3000/contentvec/
+class ContentVec:
+  # TODO: self.label_embs_concat and self.final_proj dims are hardcoded
+  # and depend on fairseq.data.dictionary Dictionary in the checkpoint.
+  # This param can't yet be loaded since there is no pickle for it. See with DEBUG=2.
+  # This means that the model only works with the given weights (they are the same for every svc model)
```

Comment:
In the original [SVC](https://github.com/svc-develop-team/so-vits-svc) they didn't actually have to reimplement the ContentVec model architecture themselves, since fairseq serializes the code for the [original model architecture](https://github.com/auspicious3000/contentvec/blob/main/contentvec/models/hubert/contentvec.py) to the checkpoint. The model + weights can then simply be ensembled from the checkpoint like [so](https://github.com/svc-develop-team/so-vits-svc/blob/64591bd664b1d4ffe88013fd2a7b1271109d7fd9/vencoder/ContentVec256L9.py#L11).

The parameter for self.final_proj is inferred from the "dictionaries" in the checkpoint like [so](https://github.com/auspicious3000/contentvec/blob/d746688a32940f4bee410ed7c87ec9cf8ff04f74/contentvec/models/hubert/contentvec.py#L321).  And [here](https://github.com/auspicious3000/contentvec/blob/d746688a32940f4bee410ed7c87ec9cf8ff04f74/contentvec/models/hubert/contentvec.py#L332C10-L332C10) for the parameter in self.label_ems_concat. [This](https://fairseq.readthedocs.io/en/latest/_modules/fairseq/data/dictionary.html) is the class definition of the dictionary.
I inferred the hardcoded value 1, and 504 (for self.final_proj, and self.label_ems_concat respectively) from the serialized model architecture/configuration in the weights.

If we wanted to use ContentVec with any other weights than what SVC trained on, then we would need to dig deeper.